File: packages/contracts/src/schemas/note.ts
// The Note domain: status/type enums, the summary, and the Firestore note doc.
//
// Source shapes:
//   - NoteStatus / NoteType / Summary / Note  ← src/types.ts (the Firestore
//     document the web app renders) reconciled with
//     ios-native/Wassup/Models/Note.swift (the client model that reads a
//     subset of the same document).
//
// The Firestore `Note` is delivered to clients via the Firestore listener, not
// a REST body, but it is the canonical domain object every surface shares, so
// it is defined here once. src/types.ts is the superset; the Swift model reads
// a subset of these fields — where they diverge, a `// TODO(contracts):` note
// records it.
import { z } from './zod';

/**
 * `processing` is retained for back-compat with notes created pre-Phase-3.
 * The chunked pipeline maps onto: queued → chunking → transcribing →
 * summarizing → ready / error. Source: src/types.ts `NoteStatus` and
 * Note.swift `NoteStatus` (identical set).
 */
export const NoteStatus = z
  .enum([
    'recording',
    'processing',
    'queued',
    'chunking',
    'transcribing',
    'summarizing',
    'ready',
    'error',
  ])
  .openapi('NoteStatus');

/** Source: src/types.ts `NoteType` / Note.swift `NoteType` (raw values). */
export const NoteType = z
  .enum(['recording', 'import_audio', 'import_pdf', 'youtube', 'scan_text', 'online_meeting'])
  .openapi('NoteType');

/**
 * The structured summary. `keyPoints` is a Firestore-only field with no
 * Postgres column (see shared/note-edit.cjs) — optional here for that reason.
 * Source: src/types.ts `Summary` and Note.swift `Summary`.
 */
export const Summary = z
  .object({
    gist: z.string(),
    actionItems: z.array(z.string()),
    keyDecisions: z.array(z.string()),
    keyPoints: z.array(z.string()).optional(),
  })
  .openapi('Summary');

/**
 * A transcript line as mirrored into the Firestore note doc (first ~200 lines,
 * ordered by start_ms). Carries only the display string `time`; the exact
 * `start_ms` lives in Postgres and arrives via `/v1/note`. Source:
 * src/types.ts `TranscriptLine`. (Note.swift adds a positional `index` and an
 * optional `startMs` client-side; those are client concerns, not wire fields.)
 */
export const FirestoreTranscriptLine = z
  .object({
    speaker: z.string(),
    text: z.string(),
    time: z.string(),
  })
  .openapi('FirestoreTranscriptLine');

/** Chunk progress for the long-audio pipeline. Source: src/types.ts
 * `Note.progress` and Note.swift `NoteProgress`. */
export const NoteProgress = z
  .object({
    done: z.number().int(),
    total: z.number().int(),
  })
  .openapi('NoteProgress');

/**
 * The Firestore note document. Superset from src/types.ts.
 *
 * TODO(contracts): Note.swift omits `audioUrl`, `fileUrl`, `language`, `tags`
 * and reads `progress`/`diagnosticCode`; the web `src/types.ts` is the
 * superset and is used as canonical. No field is renamed. Optional fields are
 * modelled `.optional()` to match the TS `?:` — the Firestore doc genuinely
 * omits them at different lifecycle points rather than sending null.
 */
export const Note = z
  .object({
    id: z.string(),
    title: z.string(),
    workspaceId: z.string(),
    authorId: z.string(),
    status: NoteStatus,
    type: NoteType,
    audioUrl: z.string().optional(),
    fileUrl: z.string().optional(),
    sourceUrl: z.string().optional(), // YouTube or meeting link
    duration: z.number().optional(),
    language: z.string().optional(),
    wordCount: z.number().optional(),
    createdAt: z.string(),
    updatedAt: z.string(),
    summary: Summary.optional(),
    transcript: z.array(FirestoreTranscriptLine).optional(),
    transcriptTruncated: z.boolean().optional(), // full transcript lives in Postgres
    rawText: z.string().optional(), // scan / OCR results
    tags: z.array(z.string()).optional(),
    errorMessage: z.string().optional(),
    diagnosticCode: z.string().optional(),
    storagePath: z.string().optional(),
    mimeType: z.string().optional(),
    jobId: z.string().optional(),
    progress: NoteProgress.optional(),
    retryAttempt: z.number().int().optional(),
    lastProgressAt: z.string().optional(),
  })
  .openapi('Note');

export type NoteStatus = z.infer<typeof NoteStatus>;
export type NoteType = z.infer<typeof NoteType>;
export type Summary = z.infer<typeof Summary>;
export type FirestoreTranscriptLine = z.infer<typeof FirestoreTranscriptLine>;
export type NoteProgress = z.infer<typeof NoteProgress>;
export type Note = z.infer<typeof Note>;

File: apps/ios/AlgoMinutes/Models/Note.swift
import Foundation

enum NoteStatus: String, Codable, CaseIterable, Sendable {
    case recording, processing, queued, chunking, transcribing, summarizing, ready, error

    var label: String {
        switch self {
        case .recording: return "Recording"
        case .processing: return "Processing"
        case .queued: return "Queued"
        case .chunking: return "Chunking audio"
        case .transcribing: return "Transcribing"
        case .summarizing: return "Summarizing"
        case .ready: return "Ready"
        case .error: return "Error"
        }
    }

    var isInProgress: Bool {
        self != .ready && self != .error
    }
}

enum NoteType: String, Codable, Sendable {
    case recording
    case importAudio = "import_audio"
    case importPdf = "import_pdf"
    case youtube
    case scanText = "scan_text"
    case onlineMeeting = "online_meeting"
}

struct Summary: Equatable, Sendable {
    var gist: String
    var actionItems: [String]
    var keyDecisions: [String]
    var keyPoints: [String]?
}

struct TranscriptLine: Equatable, Sendable, Identifiable {
    /// Position in the transcript, and the SwiftUI identity.
    ///
    /// Identity must never be derived from content. The previous
    /// `"\(speaker)|\(time)|\(text.prefix(48))"` collided on the utterances a
    /// conversation is full of — "Yes.", "Okay.", "Mm-hm." from the same
    /// speaker at the same displayed timestamp produced identical ids, and
    /// duplicate ids in a `ForEach` are undefined behaviour in SwiftUI:
    /// dropped rows, broken diffing, scroll-position jumps. That was survivable
    /// while the transcript was capped at the 200 lines Firestore mirrors; it
    /// becomes routine at the 3-6k lines a long meeting actually has.
    ///
    /// The ordinal also survives the swap from the mirrored preview to the
    /// full transcript: both are ordered by start_ms, so line n keeps id n and
    /// SwiftUI diffs an append rather than a wholesale replacement — which is
    /// what preserves scroll position when the full fetch lands.
    let index: Int
    var speaker: String
    var text: String
    var time: String
    /// Exact offset into the audio, when it is known.
    ///
    /// Firestore mirrors only the display string ("0:42"), so lines from the
    /// live listener have nil here and a tap can at best parse that back.
    /// `/api/note` returns the real `start_ms`, which is what makes
    /// tap-to-seek land on the word rather than near it.
    var startMs: Double?

    var id: Int { index }
}

struct NoteProgress: Equatable, Sendable {
    var done: Int
    var total: Int
}

/// Mirrors the Firestore note document (`src/types.ts` Note). Dates are kept
/// as ISO-8601 strings — that's what the backend writes.
struct Note: Identifiable, Equatable, Sendable {
    var id: String
    var title: String
    var workspaceId: String
    var authorId: String
    var status: NoteStatus
    var type: NoteType
    var sourceUrl: String?
    var duration: Double?
    var wordCount: Int?
    var createdAt: String
    var updatedAt: String
    var lastProgressAt: String?
    var summary: Summary?
    var transcript: [TranscriptLine]?
    var transcriptTruncated: Bool?
    var rawText: String?
    var errorMessage: String?
    var diagnosticCode: String?
    var storagePath: String?
    var mimeType: String?
    var jobId: String?
    var progress: NoteProgress?
    var retryAttempt: Int?

    var createdAtDate: Date? { Note.parseISO(createdAt) }
    var updatedAtDate: Date? { Note.parseISO(updatedAt) }
    var lastProgressAtDate: Date? { lastProgressAt.flatMap(Note.parseISO) }

    static func parseISO(_ s: String) -> Date? {
        // Backend writes `new Date().toISOString()` — fractional seconds.
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f.date(from: s) { return d }
        f.formatOptions = [.withInternetDateTime]
        return f.date(from: s)
    }

    static func isoNow(_ date: Date = Date()) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f.string(from: date)
    }
}

// MARK: - Firestore mapping (dictionary-based; resilient to partial docs)

extension Note {
    init?(id: String, data: [String: Any]) {
        guard let workspaceId = data["workspaceId"] as? String else { return nil }
        self.id = id
        self.workspaceId = workspaceId
        self.title = data["title"] as? String ?? "Untitled"
        self.authorId = data["authorId"] as? String ?? ""
        self.status = (data["status"] as? String).flatMap(NoteStatus.init(rawValue:)) ?? .processing
        self.type = (data["type"] as? String).flatMap(NoteType.init(rawValue:)) ?? .recording
        self.sourceUrl = data["sourceUrl"] as? String
        self.duration = (data["duration"] as? NSNumber)?.doubleValue
        self.wordCount = (data["wordCount"] as? NSNumber)?.intValue
        self.createdAt = data["createdAt"] as? String ?? ""
        self.updatedAt = data["updatedAt"] as? String ?? ""
        self.lastProgressAt = data["lastProgressAt"] as? String
        if let s = data["summary"] as? [String: Any] {
            self.summary = Summary(
                gist: s["gist"] as? String ?? "",
                actionItems: s["actionItems"] as? [String] ?? [],
                keyDecisions: s["keyDecisions"] as? [String] ?? [],
                keyPoints: s["keyPoints"] as? [String]
            )
        }
        if let lines = data["transcript"] as? [[String: Any]] {
            // Firestore mirrors the first 200 lines in start_ms order, so the
            // array position is the transcript position — see TranscriptLine.index.
            self.transcript = lines.enumerated().map { offset, line in
                // The fast path embeds the speaker in the text as
                // "Speaker 1: …"; reconcile before display or the row shows a
                // "Speaker" chip above text that repeats it. See SpeakerLabel.
                let split = SpeakerLabel.split(
                    speaker: line["speaker"] as? String,
                    text: line["text"] as? String ?? ""
                )
                return TranscriptLine(
                    index: offset,
                    speaker: split.label,
                    text: split.text,
                    time: line["time"] as? String ?? ""
                )
            }
        }
        self.transcriptTruncated = data["transcriptTruncated"] as? Bool
        self.rawText = data["rawText"] as? String
        self.errorMessage = data["errorMessage"] as? String
        self.diagnosticCode = data["diagnosticCode"] as? String
        self.storagePath = data["storagePath"] as? String
        self.mimeType = data["mimeType"] as? String
        self.jobId = data["jobId"] as? String
        if let p = data["progress"] as? [String: Any],
           let done = (p["done"] as? NSNumber)?.intValue,
           let total = (p["total"] as? NSNumber)?.intValue {
            self.progress = NoteProgress(done: done, total: total)
        }
        self.retryAttempt = (data["retryAttempt"] as? NSNumber)?.intValue
    }
}
